(#895) Usage of ScalarWithFallback simplified

scratchboom opened the pull request:
Usage of ScalarWithFallback simplified

0crat commented:
Job [#1042](https://github.com/yegor256/cactoos/issues/1042) is now in scope, role is `REV`

<!-- https://www.0crat.com/footprint/C63314D6Z/d5639a1c-aa1e-4049-90a7-f48b65b0b150, version: 0.40.9, hash: ${buildNumber} -->

fabriciofx commented on src/test/java/org/cactoos/scalar/ScalarWithFallbackTest.java:
```diff
@@ -58,6 +58,22 @@ public void usesMainFunc() throws Exception {
+    public void usesMainFuncInNonIterableCtor() {
+        final String message = "Main function's result";
+        MatcherAssert.assertThat(
+            "Scalar's main function value mismatch",
```
@scratchboom WDYT change this message to something like "Can't match value of Scalar's main function"?

fabriciofx commented:
@scratchboom Please, fix the PR title.

fabriciofx commented:
@scratchboom ping

scratchboom commented on src/test/java/org/cactoos/scalar/ScalarWithFallbackTest.java:
```diff
@@ -58,6 +58,22 @@ public void usesMainFunc() throws Exception {
+    @Test
+    public void usesMainFuncInNonIterableCtor() {
+        final String message = "Main function's result";
+        MatcherAssert.assertThat(
```
Done

fabriciofx commented on src/test/java/org/cactoos/scalar/ScalarWithFallbackTest.java:
```diff
@@ -96,39 +110,39 @@ public void usesFallbackOfInterruptedException() throws Exception {
             new ScalarWithFallback<>(
                 () -> {
-                    throw new IllegalFormatWidthException(1);
+                    throw new IllegalStateException("wat is raw");
```
@scratchboom What's "wat"?

scratchboom replied:
Sorry it's just left from my experiments. This string is not necessary). Removed and pulled.

fabriciofx replied:
@scratchboom Right.

fabriciofx commented:
@scratchboom This PR seems fine. Congratulations!

fabriciofx commented:
@rultor merge

rultor commented:
> @rultor merge

@fabriciofx Thanks for your request. @llorllale Please confirm this.

scratchboom commented:
@llorllale
> you've made some changes that switched exception types around.

That's because I've encountered with "ClassDataAbstractionCouplingCheck" inspection. So I decided to refactor code to minimize the count of used classes by using only some types of Exceptions. (It's doesn't matter which exception classes to use if we test executing the fallback path of ScalarWithFallback).
>Inspection message was something like that - Class Data Abstraction Coupling is 8 (max allowed is 7) classes [Assertion, FallbackFrom, IOException, IllegalFormatWidthException, InterruptedException, IsEqual, IterableOf, ScalarWithFallback]. (ClassDataAbstractionCouplingCheck)

> substituted use of `ScalarHasValue` with `IsEqual`. Why?
As @fabriciofx suggested, I've replaced `MatcherAssert.assertThat` with `new Assertion().affirm()`.
The reason why `IsEqual` used is because `Assertion` receives `Matcher<T>`, but `MatcherAssert.assertThat` receives Matcher<Scalar<T>> in case we make assertion about Scalar.
So I've replaced ScalarHasValue (which is a Matcher<Scalar<T>>) to IsEqual (which is a Matcher<T>).

And actually I've forgot to call `.affirm()` in most places and test obviously were all green. (I've already fixed and pushed this).
I think we should have an inspection, that Assertion instantiation cannot be used without calling `.affirm()`

llorllale commented on src/test/java/org/cactoos/scalar/ScalarWithFallbackTest.java:
```diff
@@ -43,25 +41,41 @@
                 )
             ),
-            new ScalarHasValue<>(message)
+            new IsEqual<>(message)
```
@scratchboom `Assertion.affirm()` is not being called in this test. Also, keep using `ScalarHasValue` and instead change line `ScalarWithFallback` to `() -> ScalarWithFallback(`...

llorllale commented:
@scratchboom 

> ClassDataAbstractionCouplingCheck

I'm OK with suppressing this in tests.

> The reason why `IsEqual` used is because `Assertion` receives `Matcher<T>`, but `MatcherAssert.assertThat` receives `Matcher<Scalar<T>>` in case we make assertion about Scalar.
So I've replaced ScalarHasValue (which is a `Matcher<Scalar<T>>`) to IsEqual (which is a `Matcher<T>`).

You could've kept using `ScalarHasValue` without any issue. I think the problem is that you are not yet fully aware of how `Assertion` is to be used.

Anyway, left you a comment above.

llorllale commented:
@scratchboom ping

llorllale commented:
@scratchboom closing due to inactivity

0crat commented:
Code review was too long (36 days), architects (@llorllale) were penalized, see [§55](http://www.zerocracy.com/policy.html#55)

<!-- https://www.0crat.com/footprint/C63314D6Z/608d4cb4-9965-45c1-8faa-144fa4c90490, version: 0.43.2, hash: ${buildNumber} -->

0crat commented:
The job [#1042](https://github.com/yegor256/cactoos/issues/1042) is now out of scope

<!-- https://www.0crat.com/footprint/C63314D6Z/9a5a2429-1dd4-4de1-bf88-98b1e7ce2144, version: 0.43.2, hash: ${buildNumber} -->

0crat commented:
Order was finished: +15 point(s) just awarded to @fabriciofx[/z](https://www.0crat.com/u/fabriciofx)

<!-- https://www.0crat.com/footprint/C63314D6Z/b6c893d1-b83f-4f06-b860-87a029bcf1df, version: 0.43.2, hash: ${buildNumber} -->

0crat commented:
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @llorllale[/z](https://www.0crat.com/u/llorllale)

<!-- https://www.0crat.com/footprint/C63314D6Z/ccea86e9-c202-40e4-871c-f77756238e8b, version: 0.43.2, hash: ${buildNumber} -->
